pulsar-common/src/main/java/org/apache/pulsar/common/naming/TopicName.java

```diff
@@ -272,7 +272,17 @@ public static int getPartitionIndex(String topic) {
         int partitionIndex = -1;
         if (topic.contains(PARTITIONED_TOPIC_SUFFIX)) {
             try {
-                partitionIndex = Integer.parseInt(topic.substring(topic.lastIndexOf('-') + 1));
+                String idx = StringUtils.substringAfterLast(topic, PARTITIONED_TOPIC_SUFFIX);
+                partitionIndex = Integer.parseInt(idx);
+                if (partitionIndex < 0) {
+                    // for the "topic-partition--1"
+                    partitionIndex = -1;
+                    log.warn("Partition index should >=0!");
```

Comment:
Should we remove the warn logs? IMO, `getPartitionIndex` may be used to check if the topic is partitioned, if the topic is just expected to be a non-partitioned topic, the warn log is redundant.

Reply:
The constructor method depends on `private TopicName(String completeTopicName)` and it depends on this method to set the `partitionIndex`, and the `isPartitioned()` method depends on the `partitionIndex`.

Reply:
But I still don't know why is the warn logs necessary? If user called `TopicName.get("my-topic-partition-xxx")`, then he'll get a line of warn logs. `TopicName` is a class of `pulsar-common` module that we should let the caller to determine whether to log something.

Reply:
Hmmm, maybe debug level is more appropriate? What do you think

Reply:
IMO, no logs is needed here.

Logging is used for helping users know what happened while the program is running. But for the package that is designed for being called by other packages, usually named `utils` or `common`, we should use exceptions or returned values to expose the running state so that caller side can know what happened. Usually, in these utils packages, a method should not contain too many steps whose internal state is invisible to caller side so logs is needed.

You can see these packages in many places of Pulsar. For example, the `org.apache.pulsar.common.protocol.Commands` class. Only `peekMessageMetadata` and `peekStickyKey` have error logs. It's because they want to swallow the exceptions and don't want callers know anything about it. However, IMO, these designs are **bad**. It actually swallows exceptions, then adds an error log, finally returns null. Unfortunately, the caller side doesn't check null value and NPE may happen. Even if we handled the null returned value, what should we do?

```java
MessageMetadata msgMetadata = Commands.peekMessageMetadata(metadataAndPayload, subscription.toString(), -1);
if (msgMetadata == null) {
    // TODO: what logs should we print?
}
```

The good design is just to propagate the exception to caller side.

```java
try {
    MessageMetadata msgMetadata = Commands.peekMessageMetadata(metadataAndPayload, subscription.toString(), -1);
    // ...
} catch (SomeException e) {
    // TODO: print logs from caller side
}
```

You can also take `org.apache.pulsar.common.net.ServiceURI` as example, no logs is used here and `IllegalArgumentException` will be thrown for fast fail.

Let's talk back to `TopicName#getPartitionIndex`. From caller side, we can check if the returned value equals to -1 so that we can know if the topic name is a non-partitioned topic, otherwise what's the partition. Your warn logs is just for printing the reason why it's a non-partitioned topic. However, caller never cares about the reason. Otherwise, it should be designed to:

```java
    /**
     * @return partition index of the completeTopicName.
     * @throws exception if the topic is not a partitioned topic
     */
    public static int getPartitionIndex(String topic) {
```

However, let's see the caller side. For example, in `org.apache.pulsar.broker.service.Consumer`'s constructor, it's used to initialize the `partitionIdx` field.

```java
this.partitionIdx = TopicName.getPartitionIndex(topicName);
```

What if the topic is a non-partitioned topic? Should we print any logs if it is? Even if the answer is yes, we only needs to print the topic name in the logs and this behavior should not be done by `getPartitionIndex`.

Reply:
BTW, the warn logs of `catch (NumberFormatException nfe)` is redundant. For example, if a user just created a topic named `my-topic-partition-not-number`, and each time this topic is checked for whether it's a partitioned topic, a warn log will be logged though no one cares about it.

Reply:
Thanks for your explanation, but throwing an exception here may not be particularly appropriate, since this method is called when the topicname is instantiated to initialize the partionIndex. 
I agree that the logs are not printed here because it doesn't matter if it's a partitioned topic or not. However, output a debug level log here for easy debugging.

Reply:
No debug logs.

Debug logs is not intended to cover all your branches and should not be abused as well. If you add logs just to ensure that the logic goes here, for example, the `TopicName` constructor may be like:

```java
            if (!completeTopicName.contains("://")) {
                log.debug("completeTopicName '{}' doesn't contains '://'", completeTopicName);
                // ...
            }
            // ...
            if (parts.size() == 3) {
                log.debug("parts.size() is 3");
                // ...
            } else if (parts.size() == 4) {
                log.debug("parts.size() is 3");
                // ...
            } else {
                log.debug("parts.size() is {},  not 3 or 4", parts.size());
                throw new IllegalArgumentException("Invalid topic name: " + completeTopicName);
            }
```

For a util method, use IDE's breakpoint for debugging. If you preferred logs, it's okay, just remove them after unit tests are added to cover all branches that have been logged.

Reply:
And for your logs

```
log.warn("Partition index should >=0!");
```

Why **should**? When we pass a topic to `getPartitionIndex`, we should not assume it's a partitioned topic. Otherwise the method should be named like `getIndexOfPartitionedTopic`.

Reply:
OK, deleted.